### PR TITLE
Allow Indirect Field Accesses in Proto Templates

### PR DESCRIPTION
**Pull request by CoolSpy3**
**Description**
Allows use of the raw `fields` variable in proto templates.

**Related Issues**
This pull-request fixes issue #6566.

**Tasks**
Add the list of tasks of this PR.
  - [x] Allow the use of the `fields` variable in proto templates.
  - [x] Add tests for the new behavior
  - [x] ~~Update the documentation~~ ~~(By my reading of the current documentation [[lua](https://cyberbotics.com/doc/reference/lua-procedural-proto?tab-language=java#programming-facts)] [[javascript](https://cyberbotics.com/doc/reference/javascript-procedural-proto#programming-facts)], the new behavior is already described.)~~ **Document the new tag**
  - [x] Clarify proto regeneration behavior
  - [x] Update the [changelog](https://github.com/cyberbotics/webots/blob/master/docs/reference/changelog-r2024.md)

**Documentation**
[Procedural PROTO Nodes (Lua)](https://cyberbotics.com/doc/reference/lua-procedural-proto?version=DeepBlueRobotics:fix-proto-template-indirect-field-references)
[Procedural PROTO Nodes (Javascript)](https://cyberbotics.com/doc/reference/javascript-procedural-proto?version=DeepBlueRobotics:fix-proto-template-indirect-field-references)
("Programming Facts" and "PROTO Regeneration" sections)

### CODE REVIEW

**Comment by CoolSpy3** [comment]
The CI failure does appear to be reproducible on this branch and not `master`, so I'm marking this PR as a draft until I have time to investigate it.

**Comment by CoolSpy3** [comment]
It looks like the `Car` proto is being regenerated as a result of these changes, causing the controller to restart. Thus, the test supervisor thinks there's three controllers running, but it only receives completion reports from two of them. I still have to find a good workaround. The only real solution I can think of is gating the behavior behind a tag. (i.e. `# tags: fieldsObject`/`# tags: indirectFieldAccess` [open to suggestions on the name]). However, if someone else has another idea, feel free to let me know.

Side note: In general, proto robots having their controllers restarted when they're regenerated seems a little unintuitive, however, any proposed behavior I can think of has some annoying edge cases. I think the best option is just to make a note of it in the proto docs. If others feel that we should update this behavior in some way, one of us can open a new issue with the proposed behavior.

**Comment by CoolSpy3** [comment]
Note: It's hard to tell from the diff, but `WbProtoModel:219` (Now `227`) and `WbProtoModel:281` (Now `292`) are not identical to their pre-edit counterparts.

File: tests/protos/worlds/template_indirect_field_access_lua.wbt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+}
+DirectionalLight {
+  direction 0 0.5 -1
+}
+Robot {
+  translation 0.1 0 0.15
+  rotation 0 1 0 1.5708
+  children [
+    DEF PROTO_template_indirect_field_access TemplateIndirectFieldAccessLua {
+		}
```
Comment (omichel) [inline tests/protos/worlds/template_indirect_field_access_lua.wbt:28]:
```suggestion
    }
```

File: tests/protos/worlds/template_indirect_field_access.wbt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+DirectionalLight {
+  direction 0 0.5 -1
+}
+Robot {
+  translation 0.1 0 0.15
+  rotation 0 1 0 1.5708
+  children [
+    DEF PROTO_template_indirect_field_access TemplateIndirectFieldAccess {
+		}
```
Comment (omichel) [inline tests/protos/worlds/template_indirect_field_access.wbt:28]:
```suggestion
    }
```

**Review by omichel — CHANGES_REQUESTED** [review]
It looks good to me.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**